llvm/lib/Transforms/InstCombine/InstCombineLoadStoreAlloca.cpp

```diff
@@ -684,6 +687,22 @@ static Instruction *combineLoadToOperationType(InstCombinerImpl &IC,
         return &Load;
       }
     }
+
+    // Fold trunc intPtrType (load intType) to load intPtrType if the load is
+    // only used by a IntToPtr instruction.
+    if (auto *TruncUser = dyn_cast<TruncInst>(Load.user_back());
+        TruncUser && TruncUser->hasOneUser()) {
+      Type *TruncTy = TruncUser->getDestTy();
+      const DataLayout &DL = IC.getDataLayout();
+      if (auto *IntToPtrUser = dyn_cast<IntToPtrInst>(TruncUser->user_back());
+          IntToPtrUser &&
+          TruncTy == DL.getIntPtrType(IntToPtrUser->getType())) {
```

Comment:
Do we routinely see IntToPtrInst? Or are they introduced by the 20->32 bit SROA trick?

Reply:
They can be produced by a chain of events when a `{ ptr }` struct is copied through a 4-byte memcpy, that gets simplified in load/stores of i32. And then to get a pointer out of the i32 alloca, one needs a trunc+inttoptr.